lib/common/xml_io.c

```diff
@@ -0,0 +1,848 @@
+/*
+ * Copyright 2004-2024 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
```

Comment:
Separate corresponding xml_io{,_internal,_compat}.h headers

Reply:
Okay. We're not doing that with `xml_display.c` or `xml_attr.c`, so I had defaulted to not doing it for `xml_io.c`.

Reply:
Yeah I want to move everything to that, and figure we may as well follow it with new files

Reply:
`xml_display.c` could reasonably be merged into `xml_io.c` BTW. I'm fine with keeping them separate, with `xml_display.c` containing all the stuff that uses `pcmk__output_t`.

Reply:
Also -- should `xml{,_internal}.h` include `xml_io{,_internal}.h`, or should files include the `io` headers directly?

Reply:
> `xml_display.c` could reasonably be merged into `xml_io.c` BTW. I'm fine with keeping them separate, with `xml_display.c` containing all the stuff that uses `pcmk__output_t`. Either way.

Separate is good. I aim for 300-1000 lines per file as being a reasonable amount to read through at one time.

> Also -- should `xml{,_internal}.h` include `xml_io{,_internal}.h`, or should files include the `io` headers directly? We decided on the former in the context of `xml_names.h`, so that's my default.
> 
> That raises the related question of whether `xml_io.c` should include `xml_io.h`... but functions within it may depend on something in `xml.h`, so I think even then, somewhat counterintuitively, it makes sense to just include `xml.h`.
> 
> I don't think `xml_io_compat.h` needs to include `xml_compat.h` but you could make that case.

For the internal files we can go right to the end goal: just include it in crm_internal.h and everything else can include that.

For the public files, we just have to ensure that any symbol previously defined by a public header still is. So, if everything in xml_io.h is split from xml.h, xml.h should include xml_io.h.

The compat files should only be included by the header they provide compatibility before. In other words xml.h will include xml_io.h which will include xml_io_compat.h.

Separately, each header should include any other headers needed to define symbols used in that header. If xml_io_compat.h needs a symbol defined in xml_compat.h, it should include xml.h to get it. So, we do need some care to avoid recursive includes (the header guards prevent it from being infinite, but it's still bad). For that reason we may occasionally need a separate header that solely defines types needed by other headers (like sched_types.h).

Reply:
> So, if everything in xml_io.h is split from xml.h, xml.h should include xml_io.h.

We could end up with circular includes if anything in `xml_io.h` ever depends on something in `xml.h`. Currently that's not the case so it's fine. (We'll only have one function in `xml_io.h`, and I plan to deprecate it very soon.)

Reply:
> The compat files should only be included by the header they provide compatibility before. In other words xml.h will include xml_io.h which will include xml_io_compat.h.

What about `xml_io.c`'s deprecated section? Presumably it includes `xml_io_compat.h` directly, if `xml_compat.h` should not include `xml_io_compat.h`.